docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+## Signals
+
+### Signal emission
+
+A widget's LCB source code must explicitly name, and provide prototypes for, the signals that it can emit [1]:
+
+    public signal <name:id>(<params:formals>)
+
+Signal declarations must be placed in module scope.  The signal **name** must be unique in the module.
+
+Signals are explicitly emitted by a widget using an `emit` statement:
+
+    emit <name:id>(<params:expressions>)
+
+It is a run-time error to emit a signal with the wrong number or types of emission parameters [2].  The compiler may statically verify that signals are emitted with compatible parameter lists.
+
+The `emit` statement may return before all the signal handlers for the signal have finished running [3].
+
+The signal handlers for a signal get run exactly once for each emission.  All of the handlers for an emission must complete before any of the handlers for a subsequent emission begin [4].
+
+The signal handlers for a signal may be run in any order.
+
+The `emit` statement does not return a value [5].
+
+**Notes**:
+
+1. These declarations provide a convenient anchor point for documentation comments about the purpose and usage of a signal.
+
+2. In principle a widget could set properties on itself before emitting a signal.  However, when adding features to LCS previously it's been found that users prefer relevant information to be passed as parameters to messages rather than out-of-band.  Passing parameters to signal emission also helps to ensure that all of the correct setup has been performed for signal handling - for example, if a parameter is added or removed from the signal, it's easy to verify that all relevant code has been updated.
+
+3. Although, initially, signal emission _could_ be implemented by iterating over registered signal handlers and calling them within the execution of the `emit` statement, this could easily lead to arbitrarily-deep signal emission chains going backwards and forwards between LCS and LCB contexts.  By allowing signal emission to be delayed, this is avoided (the current stack can be allowed to unwind before signal dispatch is performed).  Delaying signal handling also helps to reason about the safety of deleting a widget in response to receiving a signal from it.
+
+4. Emitting a signal twice from the same handler - even with the same parameters - will cause all of the handlers for that signal to be run twice.
+
+5. This is a consequence of asynchronous signal handling.
+
+### Signal handling
+
+Signal handlers are explicitly declared in an LCB program using a `signal handler` declaration [1]:
+
+    signal handler for <name:id>(<params:formals>) is <handler:constexpr>
+
```

Comment:
@peter-b Regarding point 2 is this to deal with multiple child widgets that may emit the same signal name? One thing is it could be feasible for different child widgets to emit the same name with the same formals but for logically different reasons or at least the parent widget author may want to handle them separately. Perhaps the `<name:id>` should/could include the widget kind? Maybe a nice thing to do would be something like:

````
signal handler for [<widget kind>|*]:<name:id>(...
````